corehq/warnings.py

```diff
@@ -0,0 +1,174 @@
+import os
+import re
+import warnings
+
+
+def configure_warnings(is_testing):
+    if is_testing:
+        augment_warning_messages(is_testing)
+        if 'PYTHONWARNINGS' not in os.environ:
+            warnings.simplefilter("error")
+    if is_testing or "CCHQ_WHITELISTED_WARNINGS" in os.environ:
+        configure_deprecation_whitelist()
+
+
+def configure_deprecation_whitelist():
+    """Whitelist known deprecated code paths
+
+    Adding items here is a temporary workaround to prevent a warning
+    from causing tests to fail when the warning cannot be easily
+    avoided. There is a cost associated with adding items to this list
+    (the warnings module does a linear search through the list with
+    regex matching each time a warning is triggered), so it is best to
+    avoid it if possible.
+    """
+    from django.utils.deprecation import RemovedInDjango41Warning
+    from sqlalchemy.exc import SAWarning
+
+    # warnings that may be resolved with a library upgrade
+    config_apps = "|".join(re.escape(app) for app in [
+        "captcha",
+        "django_celery_results",
+        "oauth2_provider",
+        "statici18n",
+        "two_factor",
+    ])
+    default_config_message = re.compile(fr"'.*({config_apps})' defines default_app_config")
+    whitelist("captcha.fields", "ugettext_lazy() is deprecated")
+    whitelist("celery", "'collections.abc'")
+    whitelist("couchdbkit.schema.properties", "'collections.abc'")
+    whitelist("django.apps", default_config_message, RemovedInDjango41Warning)
+    whitelist("django_celery_results", "ugettext_lazy() is deprecated")
+    whitelist("django_digest", "The 'warn' method is deprecated")
+    whitelist("django_otp.plugins", "django.conf.urls.url() is deprecated")
+    whitelist("kombu.utils.functional", "'collections.abc'")
+    whitelist("logentry_admin.admin", "ugettext_lazy() is deprecated")
+    whitelist("nose.importer", "the imp module is deprecated")
+    whitelist("nose.util", "inspect.getargspec() is deprecated")
+    whitelist("tastypie", "django.conf.urls.url() is deprecated")
+    whitelist("tastypie", "request.is_ajax() is deprecated")
+
+    # warnings that can be resolved with HQ code changes
+    whitelist("", "json_response is deprecated.  Use django.http.JsonResponse instead.")
+    whitelist("", "property_match are deprecated. Use boolean_expression instead.")
+    whitelist("corehq.util.validation", "metaschema specified by $schema was not found")
+
+    # other, resolution not obvious
+    ignored_predicates = re.compile(r"^Predicate of partial index .* ignored during reflection")
+    skipped_reflections = re.compile(
+        "^Skipped unsupported reflection of expression-based index form_processor_xformattachmentsql_blobmeta_key")
+    whitelist("sqlalchemy.", ignored_predicates, SAWarning)
+    whitelist("sqlalchemy.", skipped_reflections, SAWarning)
+    whitelist("unittest.case", "TestResult has no addExpectedFailure method", RuntimeWarning)
```

Comment:
If possible/practical, I'd love to see this as a static list defined outside of this function, and then having this function simply iterate over that static list and whitelist each item. The benefit I see of that is that if we expect a large-ish group of developers who are only peripherally aware of the underlying plumbing to be adding and removing deprecation warnings from the list, it seems safer and more straightforward to ask them to edit a static list than to edit the contents of a function.

Reply:
https://github.com/dimagi/commcare-hq/pull/31482/commits/3ef6d2f1620488b1936106672651ae86674d5742